research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+    N generations.
+After training, validation can be preformed on the holdout data. Besides, given
+two models, evaluation can be applied to choose a stronger model.
+The training pipeline consists of multiple RL learning iterations to achieve
+better models.
+"""
+
+# pylint: disable=g-bad-import-order
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import random
+import sys
+import shutil
+import socket
+import time
+from tqdm import tqdm
+
+import tensorflow as tf
+
+import dual_net
+import evaluation
+import model_utils
+import preprocessing
+import selfplay_mcts
+import utils
+
+
+def _ensure_dir_exists(directory):
+  """Check if directory exists. If not, create it.
+
+  Args:
+    directory: A given directory
+  """
+  if directory.startswith('gs://'):  # for cloud logging
+    return
+  tf.gfile.MakeDirs(directory)
+
+
+def bootstrap(estimator_working_dir, models_dir, params):
+  """Initialize the model with random weights.
+
+  Args:
+    estimator_working_dir: tf.estimator working directory.
+    models_dir: Where to export the first bootstrapped generation.
+    params: Hyper params of the model.
+  """
+  bootstrap_name = model_utils.generate(0)
+  bootstrap_model_path = os.path.join(models_dir, bootstrap_name)
+
+  _ensure_dir_exists(estimator_working_dir)
+  _ensure_dir_exists(os.path.dirname(bootstrap_model_path))
+
+  print('Bootstrapping with working dir %s\n Model 0 exported to %s' %
+        (estimator_working_dir, bootstrap_model_path))
+  dual_net.bootstrap(estimator_working_dir, params)
+  dual_net.export_model(estimator_working_dir, bootstrap_model_path)
+
+
+def selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
+             params):
+  """Perform selfplay with a specific model.
+
+  Args:
+    model_name: The name of the model used for selfplay.
+    models_dir: The path to the network model files.
+    selfplay_dir: Where to write the games'='data/selfplay/.
+    holdout_dir: Where to write the hold out data'='data/holdout/.
+    sgf_dir: Where to write the sgfs"="sgf/.
+    params: Hyper params of the model.
+  """
+  print('Playing a game with model %s' % model_name)
+  model_path = os.path.join(models_dir, model_name)
+  output_dir = os.path.join(selfplay_dir, model_name)
+  holdout_dir = os.path.join(holdout_dir, model_name)
+  clean_sgf = os.path.join(sgf_dir, model_name, 'clean')
+  full_sgf = os.path.join(sgf_dir, model_name, 'full')
+
+  _ensure_dir_exists(output_dir)
+  _ensure_dir_exists(holdout_dir)
+  _ensure_dir_exists(clean_sgf)
+  _ensure_dir_exists(full_sgf)
+
+  with utils.logged_timer('Loading weights from %s ... ' % model_path):
+    network = dual_net.DualNetwork(model_path, params)
+
+  with utils.logged_timer('Playing game'):
+    player = selfplay_mcts.play(
+        network, params.readouts, params.resign_threshold, params.verbose)
```

Comment:
Should this be selfplay readouts? As opposed evaluation of other readouts.

Reply:
Already rename it. Thanks!